storage: fix incorrect mount association

RaduBerinde opened the pull request:
We use `filepath.Rel` to associate a data directory with a mountpoint,
but the code incorrectly tolerates a result which starts with `../`.
This commit adds a check for this case.

Fixes: #137021

Release note (bug fix): Fixed bug that causes an incorrect filesystem
to be logged as part of the store information.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137024)
<!-- Reviewable:end -->

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137024#-:-ODhNJrM-zeQiTbyCuny:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
storage: fix incorrect mount association

We use `filepath.Rel` to associate a data directory with a mountpoint,
but the code incorrectly tolerates a result which starts with `../`.
This commit adds a check for this case.

Fixes: #137021

Release note (bug fix): Fixed bug that causes an incorrect filesystem
to be logged as part of the store information.

RaduBerinde commented:
Test failure is #136592

bors r+
